### PR TITLE
EDM-373: Spec Manager Test Coverage

### PR DESCRIPTION
**Pull request by DakCrowder**
Continued enhancements for the spec manager tests.  Adds new defined/shared error types and adds tests for the following functionality:
- Reading / Writing spec files
- Upgrade
- IsOSUpdate
- CheckOsReconciliation

### CODE REVIEW

**Comment by DakCrowder** [comment]
@hexfusion Let me know if the added errors / modified handling is inline with what you had in mind with your comments on https://github.com/flightctl/flightctl/pull/520!

File: internal/agent/device/spec/spec_test.go [hunk 1]
```diff
@@ -160,7 +162,7 @@ func TestInitialize(t *testing.T) {
 		mockReadWriter.EXPECT().WriteFile(gomock.Any(), gomock.Any(), gomock.Any()).Return(writeErr)
 
 		err := s.Initialize()
-		require.ErrorContains(err, "writing rollback rendered spec:")
+		require.ErrorIs(err, ErrWritingRenderedSpec)
```
Comment (hexfusion) [inline internal/agent/device/spec/spec_test.go:165]: Yes :)

File: internal/agent/device/spec/spec_test.go [hunk 2]
```diff
@@ -215,11 +217,354 @@ func TestEnsure(t *testing.T) {
...
+	filePath := "test/path/spec.json"
+
+	t.Run("error when the file does not exist", func(t *testing.T) {
+		mockReader.EXPECT().ReadFile(filePath).Return(nil, os.ErrNotExist)
+
+		_, err := readRenderedSpecFromFile(mockReader, filePath)
+		require.ErrorIs(err, ErrMissingRenderedSpec)
+	})
+
+	t.Run("error reading file returns an error containing the file path", func(t *testing.T) {
```
Comment (hexfusion) [inline internal/agent/device/spec/spec_test.go:265]: not sure this test name is accurate also we want to keep these as concise as possible.

File: internal/agent/device/spec/spec_test.go [hunk 3]
```diff
@@ -215,11 +217,354 @@ func TestEnsure(t *testing.T) {
...
+		mockReader.EXPECT().ReadFile(filePath).Return(nil, errors.New("cannot read"))
+
+		_, err := readRenderedSpecFromFile(mockReader, filePath)
+		require.ErrorIs(err, ErrReadingRenderedSpec)
+	})
+
+	t.Run("error when the file is not a valid spec", func(t *testing.T) {
+		invalidSpec := []byte("Not json data for a spec")
```
Comment (hexfusion) [inline internal/agent/device/spec/spec_test.go:273]: good test

File: internal/agent/device/spec/spec_test.go [hunk 4]
```diff
@@ -215,11 +217,354 @@ func TestEnsure(t *testing.T) {
...
+	defer ctrl.Finish()
+
+	mockReadWriter := fileio.NewMockReadWriter(ctrl)
+
+	s := &SpecManager{
+		log:              log.NewPrefixLogger("test"),
+		deviceReadWriter: mockReadWriter,
+	}
+
+	t.Run("bubbles up errors from readRenderedSpecFromFile", func(t *testing.T) {
```
Comment (hexfusion) [inline internal/agent/device/spec/spec_test.go:232]: lets make this description more concise and higher level. maybe

```
ensure proper error handling on read failure
```

File: internal/agent/device/spec/spec_test.go (hunk 2 above)
Comment (DakCrowder) [reply internal/agent/device/spec/spec_test.go:265]: ahh this was leftoever from earlier testing without the shared error types, will update

**Review by hexfusion — APPROVED** [review]

**Pull request merged**